### ISSUE TITLE
[M3] Add MLflow experiment tracking and reproducible run metadata

### ISSUE THREAD

**Issue by wisoums** [wisoums/scalerl#17]
Make MLflow the canonical experiment-tracking layer for ScaleRL **before controller tuning and deep-RL benchmark training begins**.

This issue adds reusable experiment/run logging infrastructure only. It does not train DQN/PPO or perform benchmark tuning.

## Dependencies / sequencing

Start from latest `main` after #18 is merged/finalized.

#18 owns benchmark membership. #17 must consume workload IDs/splits/version from the benchmark API rather than duplicating them.

#37 City View follows immediately after #17.

## Dependency / isolation

Use the existing optional dependency:

```toml
mlops = ["mlflow>=3.16,<4"]
```

Requirements:
- importing core `scalerl`, workloads, environment, controllers, or benchmarks must not require MLflow;
- MLflow-specific code should live in a focused optional module/package such as `scalerl.mlops`;
- if a user calls MLflow functionality without the optional dependency installed, fail with a clear install hint;
- do not add Postgres/MinIO/Docker here (#44);
- do not require a running server for unit tests.

For CI in this issue, it is acceptable to make the existing test job install `.[dev,mlops]` so MLflow tests actually execute. Do not add the broader Docker/SB3/remote-server pipeline owned by #45.

## Tracking URI

Honor standard MLflow configuration.

Do not hard-code a server URL.

Support:
- local SQLite/file-backed developer tracking;
- HTTP tracking server through `MLFLOW_TRACKING_URI`;
- the later #44 containerized tracking server without code changes.

A helper may accept an explicit tracking URI override, but environment/default MLflow configuration should work normally.

## Experiment/run model

Provide a small typed run contract, using existing Pydantic/project conventions where useful.

Every tracked run must identify at least:

### Identity / lineage
- run kind: `train`, `tune`, or `evaluate`;
- algorithm/controller name;
- Git commit SHA;
- ScaleRL version;
- Python version;
- relevant package versions;
- benchmark version when applicable;
- workload ID;
- workload split: train / validation / test;
- training seed and/or evaluation seed(s).

### Exact simulator configuration
Record the complete resolved `SimulatorConfig`, including:
- episode duration;
- control interval;
- min/initial/max replicas;
- startup delay;
- service capacity RPS;
- cost per replica-hour;
- SLA target.

The exact nested resolved config should be logged as a JSON artifact. Important scalar fields may also be flattened into MLflow params/tags for search/filtering.

### Simulator-config provenance

Record **how the simulator configuration was chosen**, not only its values.

At minimum support metadata conceptually equivalent to:
- `simulator_config_source`: e.g. `default`, `predeclared`, `calibrated_train_validation`;
- optional calibration workload IDs / note.

This is especially important for Azure.

The validated Azure v1 windows have low demand relative to the current default capacity:
- training means roughly 0.45–2.29 RPS;
- training/validation peaks are at most 4.9 RPS;
- default `service_capacity_rps = 50` means one replica can serve the observed train/validation Azure traffic.

Therefore later experiments may choose a different **experiment configuration** so Azure autoscaling is meaningful, but:
- never normalize/rescale Azure demand;
- never choose capacity from held-out test workload statistics;
- any calibrated capacity/config must be based only on train/validation data;
- MLflow must record the resulting values and their provenance/calibration workload IDs.

#17 does **not** choose that capacity. It only makes the choice traceable.

If calibration workload IDs are logged, reject or clearly fail if they contain a benchmark `test` workload.

### Reward
Record complete `RewardWeights`.

### Environment/model compatibility
Record at least:
- observation shape;
- action-space size;
- startup delay;
- control interval;
- max replicas;
- service capacity RPS;
- benchmark/config identifier where applicable.

Expose a reusable compatibility metadata representation suitable for later DQN/PPO model artifacts.

Do not silently load/declare compatibility across different observation/action contracts later.

### Algorithm/controller hyperparameters
Allow arbitrary JSON-serializable/scalar hyperparameter metadata, for example:
- threshold values;
- cooldown;
- predictive settings;
- DQN/PPO hyperparameters;
- learning timesteps/episodes.

#17 provides logging plumbing only; it must not implement those algorithms.

## Reusable API

Avoid MLflow calls scattered across future training scripts.

Provide a small reusable API/context manager, conceptually like:

```python
spec = RunSpec(...)

with start_tracked_run(spec) as run:
    log_metrics(...)
    log_artifact_dict(...)
```

Exact names may follow project style.

The API should:
1. create/select the experiment;
2. start the MLflow run;
3. log lineage/config/software metadata;
4. log the exact resolved run/config artifact;
5. expose the run ID;
6. end as FINISHED on success;
7. remain FAILED/distinguishable when the wrapped work raises.

Do not swallow the original exception merely to mark a run failed.

## Metrics

Provide reusable logging for the system metrics used by later evaluation.

Support, where applicable:
- episodic/total reward;
- infrastructure cost;
- SLA violation count/rate;
- p95 latency summaries;
- queue metrics;
- completed/dropped requests if available;
- scaling operations/churn;
- training loss/learning statistics when supplied later.

Do not invent values that a caller did not compute.

Use MLflow metric `step` for time-series/training-series logging where appropriate.

## Artifacts

Support/log JSON-safe artifacts for:
- exact resolved run specification/config;
- simulator config;
- reward weights;
- software/version metadata;
- model compatibility metadata;
- evaluation JSON/CSV when a caller provides it;
- model/checkpoint/plot paths when a caller provides them later.

#17 does not need to produce a trained model.

## Benchmark / leakage guardrail

When a run is tagged `train` or `tune`:
- its primary workload must not belong to the benchmark `test` split;
- calibration/provenance workload IDs must not include test entries.

Final `evaluate` runs may use test only later when model/controller choices are frozen.

Do not create a generic helper that silently treats all benchmark workloads as tuning data.

## Local usage

Document a lightweight local path.

Examples may use:
- SQLite URI such as `sqlite:///mlflow.db`; or
- local MLflow server on `127.0.0.1:5000`.

The code must not require the user to start a server for unit tests.

Document standard `MLFLOW_TRACKING_URI` usage for future #44.

## Tests

Tests must be isolated, deterministic, offline, and must not contact a real MLflow service.

Use temporary directories/databases.

Cover at least:

### Dependency / import isolation
- core `import scalerl` does not transitively require MLflow;
- MLflow module works when optional extra is installed.

### Run lifecycle
- successful run becomes FINISHED;
- exception inside tracked context produces a FAILED run and propagates the exception;
- run ID is accessible.

### Metadata
Verify a run records:
- controller/algorithm;
- run kind;
- Git/version/software metadata;
- workload ID/split/benchmark version;
- simulator config;
- reward weights;
- seeds;
- observation/action metadata;
- hyperparameters.

Verify exact resolved config/software/compatibility JSON artifacts exist.

### Simulator-config provenance / leakage
- provenance is logged;
- train/validation calibration workload IDs are accepted;
- a held-out test ID used as a calibration source is rejected;
- train/tune run directly targeting a test workload is rejected;
- evaluate run may represent a test workload.

### Metrics
- scalar metric logging;
- repeated metric logging with steps;
- no fabricated missing metrics.

### Tracking configuration
- isolated local SQLite/file tracking works;
- explicit/configured tracking URI is respected;
- no network service required.

### Package behavior
- optional MLflow code does not break wheel/core import behavior.

## Documentation

Update `docs/MLOPS.md` with the actual implemented API and local usage.

Update `docs/EXPERIMENTS.md` only as needed to state:
- run IDs are required for promoted results;
- simulator configuration and provenance are tracked;
- Azure capacity/config calibration may use train/validation only;
- held-out Azure test statistics may not drive capacity/reward/controller/model choices.

Keep #37 responsibility clear:
- MLflow = run lineage/metrics/artifacts;
- City View = human-readable simulation visualization;
- dashboard is not a second experiment database.

## Reporting rule

Any number promoted to README/final benchmark results must be traceable to:
- an MLflow run ID; or
- a documented aggregate of MLflow run IDs.

## Scope out

Do **not** implement:
- threshold tuning/cooldown (#13);
- predictive baseline (#14);
- DQN (#15);
- PPO (#16);
- benchmark membership (#18);
- multi-seed aggregation (#19);
- reward ablation (#20);
- City View (#37);
- Docker/Postgres/MinIO (#44);
- full ML CI pipeline (#45);
- final Azure evaluation (#46).

## Acceptance criteria

- [ ] MLflow remains an optional dependency of the core package
- [ ] reusable typed run/logging helpers exist
- [ ] tracking URI is configurable and not hard-coded
- [ ] successful/failed run lifecycle is correctly recorded
- [ ] Git/software/benchmark/workload/seed metadata is recorded
- [ ] complete simulator config + reward weights are logged
- [ ] simulator-config provenance is logged
- [ ] calibration provenance cannot reference held-out test workloads
- [ ] train/tune runs cannot target held-out test workloads
- [ ] observation/action compatibility metadata is recorded
- [ ] hyperparameters and supplied system metrics can be logged
- [ ] exact resolved JSON artifacts are logged
- [ ] local isolated tracking works without Docker/network
- [ ] CI genuinely executes MLflow tests
- [ ] docs reflect implemented behavior and Azure calibration guardrail
- [ ] Ruff, Ruff format check, mypy, and pytest pass
